### PR TITLE
Fix joinpath on Windows and improve joinpath on linux

### LINKED ISSUE #33475: joinpath bug on Windows for relative folders
```julia
 julia> joinpath("C:\\a\\b", "c:c\\e")
"c:c\\e"

```

Expected

```julia
 julia> joinpath("C:\\a\\b", "c:c\\e")
"c:\\a\\b\\c\\e"

```
Since `c:c\\e` is a relative folder in the drive `c`:, which is the same as the  `C:`,  because drive letters are case-insensitive.

### PR DESCRIPTION
**Pull request by musm**
Fix https://github.com/JuliaLang/julia/issues/33475  and https://github.com/JuliaLang/julia/issues/33455

### CODE REVIEW

**Comment by musm** [comment]
these errors are complete mystery to me. Everything works perfectly locally.

File: base/path.jl [hunk 1]
```diff
@@ -259,20 +321,7 @@ julia> joinpath("/home/myuser", "example.jl")
 "/home/myuser/example.jl"
 ```
 """
-joinpath(a::AbstractString, b::AbstractString, c::AbstractString...) = joinpath(joinpath(a,b), c...)
-joinpath(a::AbstractString) = a
-
-function joinpath(a::String, b::String)
```
Comment (musm) [inline base/path.jl]: This code is very kludgy, considering how straightforward the code is on non-Windows systems, which is  especially true since we need  a custom Windows implementation anyways.

The improved code is about 20x faster in common cases, and allocates far fewer.

File: base/path.jl [hunk 2]
```diff
@@ -246,8 +246,70 @@ function splitpath(p::String)
...
+    end
+
+    return result_drive * result_path
+end
+
+else
+
+function joinpath(path::AbstractString, paths::AbstractString...)
+    for p in paths
+        if !isempty(p) && p[1] == '/'
```
Comment (musm) [inline base/path.jl]: `isabspath(p)`

File: base/path.jl [hunk 3]
```diff
@@ -246,8 +246,70 @@ function splitpath(p::String)
...
+                result_drive = p_drive
+                result_path = p_path
+                continue
+            end
+            # same drive in different case
+            result_drive = p_drive
+        end
+
+        # second path is relative to the first
+        if !isempty(result_path) && result_path[end] ∉ ('\\', '/')
```
Comment (musm) [inline base/path.jl]: Should I use  `last` and `first` instead of `x[end]` and `x[1]`. I'm not sure if there is a convention or advantage to one or the other when `x` is a string?

File: base/path.jl (hunk 1 above)
Comment (StefanKarpinski) [reply base/path.jl]: Calling the code that someone else wrote kludgy when it is very likely that person who you'll need to review your code seems like a bad way to approach getting changes reviewed. There's no need to shit on the existing code in order to make improvements.

Comment (musm) [reply base/path.jl]: I deeply apologize to anyone if this came across offensive, as it was definitely not  intention to do so.   I just found the previous logic  hard to follow and confusing, which is what I mean when I said "kludgy"; I totally didn't mean anything malicious (poor comment on my behalf).

File: base/path.jl (hunk 3 above)
Comment (StefanKarpinski) [reply base/path.jl]: They should be equivalent so whichever seems clearer to you works.

File: base/path.jl [hunk 4]
```diff
@@ -246,8 +246,70 @@ function splitpath(p::String)
     return out
 end
 
+joinpath(path::AbstractString) = path
```
Comment (StefanKarpinski) [inline base/path.jl]: Might as well make the function type stable while you're at it and always return `String`.

**Comment by StefanKarpinski** [comment]
If you're concerned about efficiency and not creating lots of intermediary strings—which you seem to be—then you may want to use the `sprint` function to construct the output strings incrementally via printing to an IOBuffer. Of course, I don't think efficiency matters much for operations on file paths, but that would be the more efficient approach here.

File: base/path.jl [hunk 5]
```diff
@@ -246,8 +246,70 @@ function splitpath(p::String)
...
+            continue
+        elseif !isempty(p_drive) && p_drive != result_drive
+            if lowercase(p_drive) != lowercase(result_drive)
+                # different drives, ignore the first path entirely
+                result_drive = p_drive
+                result_path = p_path
+                continue
+            end
+            # same drive in different case
+            result_drive = p_drive
```
Comment (StefanKarpinski) [inline base/path.jl]: Does it make sense to normalize the drive name capitalization?

File: base/path.jl (hunk 3 above)
Comment (StefanKarpinski) [reply base/path.jl]: I think what you're doing now is clearer and how I would write it.

File: base/path.jl (hunk 5 above)
Comment (StefanKarpinski) [reply base/path.jl]: Are UNC drives also case insensitive?

Comment (musm) [reply base/path.jl]: > Does it make sense to normalize the drive name capitalization?

Don't think so. We should just preserve the input capitalization. 

> Are UNC drives also case insensitive?

In this case it doesn't matter since if the user specifies a UNC path it is absolute.

**Comment by musm** [comment]
@StefanKarpinski: Thank you very much for the detailed review. I went ahead and also removed the **troublesome** `pathsep` function, which apparently was only used by `joinpath`. As mentioned in  https://github.com/JuliaLang/julia/issues/33455,  this function behaves incorrectly  due to an odd compiler(?) bug. (In particular, it gives the wrong answer when used in `joinpath`, but if you recompile the `joinpath` function it then gives the correct output). It was my mistake to wrongly assume this bug was due to the recursive nature of the function, which is simply not true and is my fault.

The current implementation of `joinpath` is this correct and irrelevant to the issue in https://github.com/JuliaLang/julia/issues/33455, *but* there is still a bug due the issue in https://github.com/JuliaLang/julia/issues/33475, which would require using `lowercase`, in the current implementation.

Hence, it is probably possible to keep the current implementation and check drive letter lowercase issues on Windows with an additional if-clause. It's also possible, if there is a preference, to use the implementation in this PR. Both options I am comfortable with. The advantage of the current implementation is speed, but as you mention it probably doesn't matter for such a non-critical path API function. The advantage of the existing `joinpath` is conciseness and elegance (I haven't fully investigated what it would take to fix the drive casing issue but as mentioned I don't think too much effort is needed to fix https://github.com/JuliaLang/julia/issues/33475)

File: test/path.jl [hunk 6]
```diff
@@ -31,15 +31,33 @@
...
         @test joinpath(S("foo"), S(homedir())) == homedir()
         @test joinpath(S(abspath("foo")), S(homedir())) == homedir()
 
         if Sys.iswindows()
             @test joinpath(S("foo"),S("bar:baz")) == "bar:baz"
             @test joinpath(S("C:"),S("foo"),S("D:"),S("bar")) == "D:bar"
             @test joinpath(S("C:"),S("foo"),S("D:bar"),S("baz")) == "D:bar$(sep)baz"
+
+            # relative folders and case-insensitive drive letters
+            @test joinpath(S("C:\\a\\b"), S("c:c\\e")) == "c:\\a\\b\\c\\e"
```
Comment (StefanKarpinski) [inline test/path.jl]: Should the earlier or later capitalization of the drive name be kept?

Comment (musm) [reply test/path.jl]: The preference is to keep the later capitalization. This follows the general rule to prefer the  path  of `p2`  in `joinpath(p1,p2)`, or if `p2` is absolute to override `p1` completely.

Comment (StefanKarpinski) [reply test/path.jl]: What does Python do here? I would think that in the case where `a` is absolute and `b` is relative, if they are joinable (i.e. the drives are compatible) one would want `joinpath(a, b)` to start with `a` and end with `b`.

Comment (musm) [reply test/path.jl]: > What does Python do here? I


Python keeps the later capitalization.  But it doesn't really matter, since it is case-insentive.

Comment (StefanKarpinski) [reply test/path.jl]: That's interesting. I was thinking that this might affect the definition of `relpath` but that doesn't seem to depend on `joinpath` at all in its logic, although I suspect it probably does have a bug in the case of drive letters with case mismatches.

Comment (musm) [reply test/path.jl]: I see, honestly I haven't had a close look at `relpath` to tell how this affects `relpath`. My general thinking is that it would be a good idea to audit all of our FileSystem API, to check how we do with UNC paths. We also don't support extended paths, which is another good extension in the future.

**Comment by musm** [comment]
@StefanKarpinski  It seems that this function is also a candidate for eager conversion. In other words I should define 
```
    joinpath(path::AbstractString, paths::AbstractString...) = _joinpath(String(path), String.(paths)...)

```
for performance reasons.

**Comment by StefanKarpinski** [comment]
Yes, this would be a good case for early conversion.

**Comment by musm** [comment]
Great, and I also confirmed locally that this early conversion does help a lot with performance and reducing allocations.

**Comment by StefanKarpinski** [comment]
Excellent. It should also make things easier on the compiler which needs to specialize less code for "weird string types".

**Comment by StefanKarpinski** [comment]
This seems good to me, I just need to look into https://github.com/JuliaLang/julia/issues/33455#issuecomment-540113905 and understand what's going on there. If I forget, please ping me to remind me about this because we definitely want this.

**Comment by musm** [comment]
NP. And thanks for taking a look at this.

Yes the current bug in `joinpath` needs closer inspection, because the compiler is messing up the `pathsep` function.

**Comment by StefanKarpinski** [comment]
Do you think that if we weren't hitting that bug then the current `joinpath` implementation would work? If so, that alleviates my concern about the recursive identity and makes this PR a combination of compiler bug workaround and performance optimization. Is there any other functional behavior change that this PR makes?

**Comment by musm** [comment]
If we weren't hitting that bug then the current implementation works for https://github.com/JuliaLang/julia/issues/33455 but not https://github.com/JuliaLang/julia/issues/33475 (it would require more tweaks to get it to fix that later bug)

So this PR works around the compiler bug by performance optimization and in the Windows case it also fixes https://github.com/JuliaLang/julia/issues/33475 by testing another branch (lowercase)

**Comment by StefanKarpinski** [comment]
Ok, thanks for the explanation and reassurances. That makes me feel pretty good about this. I will leave it open for now while I look into the compiler issue.

**Comment by musm** [comment]
Maybe the compiler got smarter, or more likely my benchmarks might have been sloppy done,  but now I see no performance difference between the `AbstractString` implementation and the one that does eager conversion to `String`.

**Comment by StefanKarpinski** [comment]
The only thing I would change at this point is keeping the earlier drive name when they have a case mismatch, since I would expect `joinpath("C:\\foo", "c:baz")` to produce `C:\\foo\\baz`.

File: base/path.jl [hunk 7]
```diff
@@ -246,8 +238,69 @@ function splitpath(p::String)
...
+            result_path = p_path
+            continue
+        elseif !isempty(p_drive) && p_drive != result_drive
+            if lowercase(p_drive) != lowercase(result_drive)
+                # different drives, ignore the first path entirely
+                result_drive = p_drive
+                result_path = p_path
+                continue
+            end
+            result_drive = p_drive # same drive in different case
```
Comment (StefanKarpinski) [inline base/path.jl]:
```suggestion
```
Just delete this line, I think.

File: test/path.jl (hunk 6 above)
Comment (StefanKarpinski) [reply test/path.jl]:
```suggestion
            @test joinpath(S("C:\\a\\b"), S("c:c\\e")) == "C:\\a\\b\\c\\e"
```
And change this test.

**Review by StefanKarpinski — CHANGES_REQUESTED** [review]

**Comment by musm** [comment]
@StefanKarpinski  I  see where you are coming from. At the same time, I think the current behavior is preferred, since in `joinpath(a,b)`, the argument `b` has  "preference". Thus I'm inclined to keep the current behavior if that's ok.

**Comment by StefanKarpinski** [comment]
I'm not sure I agree that `b` has preference. If has preference only if `b` is absolute, in which case, yes, we should take the drive value from `b`. If `b` is not absolute then `a` is the start of the result and `b` is the rest, in which case it makes much more sense to keep the drive from `a`.

**Comment by musm** [comment]
Updated

**Review by StefanKarpinski — APPROVED** [review]
